ENCD-5171-download-fastq-files

forresttanaka opened the pull request:
* I noticed JSDoc doesn't really adequately handle React hooks, so I'm trying out an alternative style of my own, but hopefully some de facto documentation standard comes out.
* I completely rewrote the DownloadButton component to now offer two kinds of dropdown buttons
    * Immediate: The button includes the chevron and executes an action the moment you choose an option from the dropdown menu.
    * Selected: “Github style” — The button and chevron are separate buttons combined, and selecting an option from the dropdown changes the label of the button, and clicking the button then executes the action with the chosen option.
* We have two `<DropdownButton>` uses outside of carts that I converted to use the new `<DropdownButton.Immediate>` component but haven’t tested it because I’m fairly confident that code can never execute. One is in ExperimentSeries when visualizable files exist, but the `visualizable` property is not implemented on ExperimentSeries objects. The other is in edit forms but I could not find a situation that would make that button render.
* In cart/batch_download.js I moved the `batchDownload` function out of `CartBatchDownloadComponent` because I converted to React hooks and didn’t want such a large function allocated on every execution of `CartBatchDownloadComponent.`
* `<DropdownButton.Selected>` is part of a new reusable component to manage the new Download button.
* In view_controls.js, I split the component to display the Download button and modal for non-cart pages into two components — one for the button and one for the modal. This allows cart batch download to use just the modal without also including the button, as the Download button for the cart isn’t a simple button anymore, while non-cart pages do still use a simple button. This modal had to change from plug-and-play into a managed modal.
* For the back end, I removed the “visualizable=true” query string parameter and replaced it with an “option=” parameter. The possible values are “visualizable” and “raw.” The “visualizable” option is just like the old “visualizable=true” parameter, while “option=raw” only downloads files that have no assembly.
* I think this is our first-ever use of custom React hooks, with `useDropdownButton` in button.js. It’s quite a bit like the old mixins so that multiple components can embed the same functionality. The unusual thing I did with this one was to return an object so that I don’t have to return so many array items, which is what you see with most custom React hooks.

forresttanaka committed:
ENCD-5171-dl-cart-fastq

forresttanaka commented:
Updated demo: ~https://encd-5171-dl-cart-fastq-pr1-fytanaka.demo.encodedcc.org/~

emro commented on src/encoded/static/scss/encoded/modules/_cart.scss:
```diff
@@ -1097,3 +1093,41 @@ $file-type-colors: (
+
+            &:hover {
+                color: #fff;
+                
```
extra space?

emro reviewed (dismissed):
I looked through everything carefully and it all looks terrific! I like the implementation too, the new drop-down button type, the way that the facets are grayed out on the raw data tab, etc. I wish I'd had one more day to come back to this because I sometimes notice new things that way but it's unlikely!

forresttanaka committed:
First fixes from Idan’s QA

No longer evaluate no-assembly files for facets. Rename download button.

forresttanaka committed:
Remove file count

Can’t really tell how many we’re actually going to download anymore.

forresttanaka commented:
New demo: https://encd-5171-dl-cart-fastq-pr2-fytanaka.demo.encodedcc.org/

forresttanaka committed:
Disable Download button if no files downloadable

This now applies to all three different download types.

forresttanaka committed:
Fix up BDD tests

forresttanaka commented:
For the updates since the last code review:

* To disable the Download button if no files can be downloaded, I now pass three different counts in `fileCounts` to be used depending on which dropdown option the user selected.
* Changed CartBatchDownloadComponent to have downloadType be a state instead of a ref because I now use it within the returned JSX.
* I added a red banner to the batch download modal explaining why the user can't download files.

Updated demo:
https://encd-5171-dl-cart-fastq-pr2-fytanaka.demo.encodedcc.org/
